src/vs/workbench/services/untitled/common/untitledTextEditorModel.ts

```diff
@@ -368,6 +389,19 @@ export class UntitledTextEditorModel extends BaseTextEditorModel implements IUnt
 
 		// Emit as general content change event
 		this._onDidChangeContent.fire();
+
+		this.detectLanguageIfEnabled();
+	}
+
+	@debounce(600)
+	private async detectLanguageIfEnabled() {
+		if (this.hasModeSetExplicitly || !this.languageDetectionService.isEnabledForMode(this.getMode() ?? '')) {
+			return;
+		}
+
+		const lang = await this.languageDetectionService.detectLanguage(this.resource);
+		if (!lang) { return; }
+		this.setModeInternal(lang);
```

Comment:
Is it possible to run into race conditions here? E.g.:
- the model may be disposed already by the time this returns (e.g. editor closed)
- the model might have been changed and another language detection run occurs, should we wire through some kind of cancellation support?

Reply:
> the model may be disposed already by the time this returns (e.g. editor closed)

I added a check around setting the mode that checks if it's been disposed.

> the model might have been changed and another language detection run occurs, should we wire through some kind of cancellation support?

Possible. Unfortunately, Tensorflow doesn't support cancellation natively... the best we can do is `Promise.race` with a token being cancelled. Is there prior art somewhere on that? I'd like to save this for later but I'll open an issue about it if that's ok with you.

Reply:
Yeah one ugly bit of the `@debounce` annotation is that we cannot clear the timeout when the model is disposed. That's why I typically prefer to manage my own `RunOnceScheduler` which is disposable.

Reply:
I had no idea about `RunOnceScheduler`! that's nice!